// ydb-sdk/build/esm/src/query/symbols.js
"use strict";
/**
 * Symbols of methods internal to the package
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.resultsetYdbColumns = exports.sessionCurrentOperation = exports.sessionTxSettings = exports.sessionTxId = exports.sessionId = exports.sessionRollbackTransaction = exports.sessionCommitTransaction = exports.sessionBeginTransaction = exports.sessionAttach = exports.sessionIsDeleted = exports.sessionDeleteOnRelease = exports.sessionIsClosing = exports.sessionIsFree = exports.sessionRelease = exports.sessionAcquire = exports.create = void 0;
exports.create = Symbol('create');
exports.sessionAcquire = Symbol('sessionAcquire');
exports.sessionRelease = Symbol('sessionRelease');
exports.sessionIsFree = Symbol('sessionIsFree');
exports.sessionIsClosing = Symbol('sessionIsClosing');
exports.sessionDeleteOnRelease = Symbol('sessionDeleteOnRelease');
exports.sessionIsDeleted = Symbol('sessionIsDeleted');
// export const sessionDelete = Symbol('sessionDelete'); // Note: Symbol named method do not support decorators
exports.sessionAttach = Symbol('sessionAttach');
exports.sessionBeginTransaction = Symbol('sessionBeginTransaction');
exports.sessionCommitTransaction = Symbol('sessionCommitTransaction');
exports.sessionRollbackTransaction = Symbol('sessionRollbackTransaction');
exports.sessionId = Symbol('sessionId');
exports.sessionTxId = Symbol('sessionTxId');
exports.sessionTxSettings = Symbol('sessionTxSettings');
exports.sessionCurrentOperation = Symbol('sessionCurrentOperation');
exports.resultsetYdbColumns = Symbol('resultsetYdbColumns');
