// ydb-sdk/build/esm/src/__tests__/e2e/query-service/rows-conversion.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const discovery_service_1 = __importDefault(require("../../../discovery/discovery-service"));
const logging_1 = require("../../../logging");
const anonymous_auth_service_1 = require("../../../credentials/anonymous-auth-service");
const constants_1 = require("../../../constants");
const query_session_pool_1 = require("../../../query/query-session-pool");
const types_1 = require("../../../types");
const DATABASE = '/local';
const ENDPOINT = 'grpcs://localhost:2136';
const TABLE_NAME = 'test_table_3';
class Row extends types_1.TypedData {
    id;
    rowTitle;
    time;
    constructor(data) {
        super(data);
        this.id = data.id;
        this.rowTitle = data.rowTitle;
        this.time = data.time;
    }
}
__decorate([
    (0, types_1.declareType)(types_1.Types.UINT64)
], Row.prototype, "id", void 0);
__decorate([
    (0, types_1.declareType)(types_1.Types.UTF8)
], Row.prototype, "rowTitle", void 0);
__decorate([
    (0, types_1.declareType)(types_1.Types.DATETIME)
], Row.prototype, "time", void 0);
describe('Rows conversion', () => {
    let discoveryService;
    let session;
    beforeAll(async () => {
        await testOnOneSessionWithoutDriver();
    });
    afterAll(async () => {
        discoveryService.destroy();
        await session.delete();
    });
    it('Ydb to native', async () => {
        await createTestTable();
        await insertCupleLinesInTestTable();
        const res = await simpleSelect(0 /* RowType.Native */);
        for await (const rs of res.resultSets) {
            expect(rs.index).toBe(0);
            expect(rs.columns[0]).toBe('id');
            expect(rs.columns[1]).toBe('rowTitle'); // as camel case
            const { value: row1 } = await rs.rows.next();
            expect(row1.id).toBe(1);
            expect(row1.rowTitle).toBe('Some title1');
            const { value: row2 } = await rs.rows.next();
            expect(row2.id).toBe(2);
            expect(row2.rowTitle).toBe('Some title2');
        }
    });
    it('Ydb to typed structure', async () => {
        await createTestTable();
        await insertCupleLinesInTestTable();
        const res = await simpleSelect(1 /* RowType.Ydb */);
        for await (const rs of res.resultSets) {
            expect(rs.columns[0].name).toBe('id');
            expect(rs.columns[1].name).toBe('row_title'); // snake case as in YDB
            expect(rs.index).toBe(0);
            const typedRows = rs.typedRows(Row);
            const { value: row1 } = await typedRows.next();
            expect(row1.id).toBe(1);
            expect(row1.rowTitle).toBe('Some title1');
            const { value: row2 } = await typedRows.next();
            expect(row2.id).toBe(2);
            expect(row2.rowTitle).toBe('Some title2');
        }
    });
    async function createTestTable() {
        await session.execute({
            text: `
                DROP TABLE IF EXISTS ${TABLE_NAME};

                CREATE TABLE ${TABLE_NAME}
                (
                    id        UInt64,
                    row_title Utf8, -- NOT NULL,
                    time      Timestamp,
                    PRIMARY KEY (id)
                );`,
        });
    }
    async function insertCupleLinesInTestTable() {
        await session.execute({
            parameters: {
                '$id1': types_1.TypedValues.uint64(1),
                '$title1': types_1.TypedValues.text('Some title1'),
                '$id2': types_1.TypedValues.uint64(2),
                '$title2': types_1.TypedValues.text('Some title2'),
                '$timestamp': types_1.TypedValues.timestamp(new Date()),
            },
            text: `
                INSERT INTO ${TABLE_NAME} (id, row_title, time)
                VALUES ($id1, $title1, $timestamp);
                INSERT INTO ${TABLE_NAME} (id, row_title, time)
                VALUES ($id2, $title2, $timestamp);
            `,
        });
        return 2;
    }
    async function simpleSelect(rowMode) {
        return await session.execute({
            rowMode,
            text: `
                SELECT *
                FROM ${TABLE_NAME};
            `,
            // concurrentResultSets: false,
        });
    }
    async function testOnOneSessionWithoutDriver() {
        const logger = (0, logging_1.getLogger)();
        const authService = new anonymous_auth_service_1.AnonymousAuthService();
        discoveryService = new discovery_service_1.default({
            endpoint: ENDPOINT,
            database: DATABASE,
            authService,
            discoveryPeriod: constants_1.ENDPOINT_DISCOVERY_PERIOD,
            logger,
        });
        await discoveryService.ready(constants_1.ENDPOINT_DISCOVERY_PERIOD);
        const sessionBuilder = new query_session_pool_1.SessionBuilder(await discoveryService.getEndpoint(), DATABASE, authService, logger);
        session = await sessionBuilder.create();
    }
});
